src/Client/Sdk/Storage/InMemoryRepository.cs

```diff
@@ -1,82 +1,27 @@
 ﻿using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Linq;
-using System.Linq.Expressions;
 
 namespace System.Net.Mqtt.Sdk.Storage
 {
 	internal class InMemoryRepository<T> : IRepository<T>
-		where T : StorageObject
+		where T : IStorageObject
 	{
-		readonly ConcurrentDictionary<string, T> elements;
+		readonly ConcurrentDictionary<string, T> elements = new ConcurrentDictionary<string, T> ();
 
-		public InMemoryRepository ()
-		{
-			elements = new ConcurrentDictionary<string, T> ();
-		}
-
-		public IEnumerable<T> GetAll (Expression<Func<T, bool>> predicate = null)
-		{
-			var result = elements.Select(x => x.Value);
-
-			if (predicate != null) {
-				result = result.Where (predicate.Compile ());
-			}
-
-			return result;
-		}
+		public IEnumerable<T> GetAll () => elements.Select (e => e.Value);
 
 		public T Get (string id)
 		{
-			var element = default (T);
-
-			elements.TryGetValue (id, out element);
+			elements.TryGetValue (id, out T element);
 
 			return element;
 		}
 
-		public T Get (Expression<Func<T, bool>> predicate)
-		{
-			return GetAll ().FirstOrDefault (predicate.Compile ());
-		}
+		public void Create (T element) => elements.TryAdd (element.Id, element);
 
-		public bool Exist (string id)
-		{
-			return Get (id) != default (T);
-		}
-
-		public bool Exist (Expression<Func<T, bool>> predicate)
-		{
-			return Get (predicate) != default (T);
-		}
+		public void Update (T element) => elements.AddOrUpdate (element.Id, element, (key, value) => element);
```

Comment:
Given this, why not a `Save` that covers both `Add` and `Update`?

Reply:
Yes, a Save instead of a Create and Update would be fine too